### ISSUE TITLE
Suggestion: Background Resource Loader 

### ISSUE THREAD

**Issue by Firemanarg** [FireDroidGameStudios/godot-project-template#3]
## Suggestion name

Background Resource Loader

## Description of suggestion

- Implement a module for FDCore to handle threaded background resource loading;
- It must include customization of loading screens, by setting them in the Project Settings. If :
    - _initial loading screen_: loading screen shown before the intro animations. If this value is `null`, the _default loading screen_ is used as _initial loading screen_;
    - _default loading screen_: loading screen shown in all game (except the loading before intro animations). This value is mandatory in Project Settings;
- It must have a loading queue, where new resources can be added to load, then all resources in load queue will be loaded.
- It must also have the following methods:
    - `FDCore.add_to_load_queue(resource_path: StringName) -> bool`: add a resource to load queue by passing a `resource_path` location. If the resource was previously loaded, if it is already queued or if a load is in progress, the resource is not added to the load queue;
    - `FDCore.get_load_queue() -> Array[StringName]`: return an array containing all the resource_paths in load queue;
    - `FDCore.clear_load_queue() -> void`: clear all resource_paths in load_queue. If a load is in progress, it does nothing and a warning is printed;
    - `FDCore.start_load() -> void`: start loading of resources in load_queue. If load is already in progress, it does nothing and a warning is printed.
    - `FDCore.abort_load() -> void`: if a loading of resources is in progress, abort it. Otherwise does nothing and a warning is printed. To prevent errors, before aborting, this method should finish loading current resource that is being loaded.
    - `FDCore.get_load_progress() -> float`: if a load of resources is in progress, this method returns the progress of the loading as a float number between 0.0 and 1.0. If no load is in progress, the method returns 1.0;
- It must also have the following signals:
    - `load_started`: emitted when the loading of resources start (by calling method `FDCore.start_load`;
    - `load_finished`: emitted when all the resources from load queue are loaded.

**Comment by Firemanarg** [comment]
# Modelling nº 1
```mermaid
# Modelling
```mermaid
---
title: Class Diagram - FDLoad (submodule of FDCore)
---
classDiagram
    class FDLoad {
        - _load_queue: Array[StringName]
        - _queue_faluires: Array[StringName]
        - _progress_array: Array
        - use_subthread: bool
        - signal started()
        - signal finished()
        - signal aborted()
        - signal progress_changed(progress: float)
        + add_to_queue(path: StringName): bool
        + get_queue(): Array[StringName]
        + get_failures(): Array[StringName]
        + retry_failures(): bool
        + clear_queue(): void
        + start(): void
        + abort(): void
        + get_progress(): float
        + get_loaded(path: StringName): 
    }

```

# Description
## Properties
- _load_queue: Array containing all resource paths to be loaded
- _queue_faluires: Array containing all resource paths from queue that failed to be loaded
- _progress_array: Array to receive progress from built-in method

## Signals
- started: Emitted when load starts successfully
- finished(has_failures): Emitted when load finishes successfully (without being aborted). First argument is a bool that is set if there was at least one failure during load
- aborted: Emitted when load finishes by abortion (manual interruption)
- progress_changed(progress): Emitted when a resource from the progress has changed. First argument is a float value between 0.0 and 1.0

## Methods
- add_to_queue(path): Add a resource path to queue and return true if it doesn't exists on the queue
- get_queue: Return the queue array
- get_failures: Return the failures array
- retry_failures: Retry to load all resources on the failures queue. If the resource fails again, it will be re-added to the failures queue
- clear_queue: Clear the load queue
- start: Begin the loading process of all resource paths from the load_queue. If a resource fails to load, it will be added to failures queue
- 
```

# Description
## Properties
- _load_queue: Array containing all resource paths to be loaded
- _queue_faluires: Array containing all resource paths from queue that failed to be loaded
- _progress_array: Array to receive progress from built-in method

## Signals
- started: Emitted when load starts successfully
- finished(has_failures): Emitted when load finishes successfully (without being aborted). First argument is a bool that is set if there was at least one failure during load
- aborted: Emitted when load finishes by abortion (manual interruption)
- progress_changed(progress): Emitted when a resource from the progress has changed. First argument is a float value between 0.0 and 1.0

## Methods
- add_to_queue(path): Add a resource path to queue and return true if it doesn't exists on the queue
- get_queue: Return the queue array
- get_failures: Return the failures array
- retry_failures: Retry to load all resources on the failures queue. If the resource fails again, it will be re-added to the failures queue
- clear_queue: Clear the load queue
- start: Begin the loading process of all resource paths from the load_queue. If a resource fails to load, it will be added to failures queue
- abort: Interrupts the loading process and emit the signal `aborted`. This method does not clear the load queue nor the failures queue
- get_progress: Return the progress of current loading. If there is no loading in progress, return 1.0
- get_loaded(path): Return the loaded resource of given path. If the resource haven't been loaded previously, it will be loaded individually (so this method works both to queue loads and individual loads without adding to queue)

**Comment by Firemanarg** [comment]
# Modelling nº 2
```mermaid
---
title: Class Diagram - FDLoad (submodule of FDCore) and FDLoadRequest (subclass of FDLoad)
---
classDiagram
    class FDLoad {
        - abort_on_failure: bool
        - _load_queue: Array[FDLoadRequest]
        - _sub_load_queue: Array[FDLoadRequest]
        - _queue_faluires: Array[FDLoadRequest]
        - _current_load_type: _LoadType
        + signal started()
        + signal finished(has_failures: bool)
        + signal aborted()
        + signal progress_changed(progress: float)
        + add_to_queue(path: String, string_type: String, use_subthread: bool, cache_mode: CacheMode): bool
        + get_queue(): PackedStringArray
        + get_failures(): PackedStringArray
        + has_failures(): bool
        + retry_failures(): bool
        + clear_queue(): void
        + start(): void
        + get_progress(): float
        + has_loaded(path: String): bool
        + get_loaded(path: String): Resource
    }
    class _LoadType {
        <<Enumeration>>
        DEFAULT,
        RETRY
    }
    class FDLoadRequest {
        - path: String
        - use_subthread: bool
        - cache_mode: CacheMode
        - string_type: String
        - _progress: Array
        + signal finished()
        + signal failed()
        + get_progress(): float
        + start_load(): void
        + is_loaded(): bool
    }
    _LoadType --> FDLoad
```
`Removed manual abort method. Now abort will be a flag. Each FDLoadRequest manages its own progress and emit a signal when finished.`
 
## Class: FDLoad
### Properties
- `abort_on_failure`: Flag to define if the loading will abort on first failure or if it will continue and storage the failures for a retry. If this flag is set, only the first failure will be added to failures queue
- `_load_queue`: Array containing all resource load requests
- `_sub_load_queue`: Sub queue to temporary storage load_queue requests when retrying failures
- `_queue_faluires`: Array containing all resource load requests from queue that failed to be loaded
- `_current_load_type`: Enum value to check if current load is a regular load (DEFAULT) or a failure retry load (RETRY)

### Signals
- `started`: Emitted when load starts successfully
- `finished`: Emitted when load finishes successfully (without being aborted). First argument is a bool that is set if there was at least one failure during load
- `aborted`: Emitted when load finishes by abortion (only if `abort_on_failure` flag is set)
- `progress_changed(progress)`: Emitted when progress change

### Methods
- `add_to_queue(path, string_type, use_subthread, cache_mode)`: Create a load request from a resource path. Next arguments are optional, with default values `""`, `true` and `ResourceLoader.CacheMode.CACHE_MODE_REUSE` respectivelly for `string_type`, `use_subthread` and `cache_mode`
- `get_queue()`: Return a PackedStringArray containing load resource paths in queue to load
- `get_failures()`: Return a PackedStringArray containing load resource paths in failures queue
- `has_failures()`: Return true if failures queue is not empty
- `retry_failures()`: Clear the load queue and add all requests from the failures queue
- `clear_queue()`: Delete all requests from `load_queue` and `failures_queue`
- `start()`: Begin the loading process of all resource requests from the load_queue. If a resource fails to load, it will be added to failures queue, and the load is aborted if the flag `abort_on_failure` is set. If failures queue contains any value before this method is called, it will be lost, since this method clears the failures queue when called
- `get_progress()`: Calculates the arithmetic average progress of all requests. Internally, this method must convert float values to int before dividing, to optimize processing. Use bitwise operations ((p << 3) + (p << 1)) to multiply by 10. After the sum of all values, divide it by the amount of requests multiplied by 100.
- `has_loaded(path)`: Return `true` if the resource on `path` has been loaded previously
- `get_loaded(path)`: Return the loaded resource of given path. If the resource haven't been loaded previously, null is returned

## Class: FDLoadRequest
### Properties
- `path`: Path of the resource to be loaded
- `use_subthread`: If false, loading process occurs in main thread, otherwise in subthread
- `cache_mode`: Cache mode of loading (check CacheMode from ResourceLoader)
- `string_type`: Optional property. If empty, the loaded resource will be handled as a generic Resource
- `_progress`: Array to receive progress from ResourceLoader

### Signals
- `finished`: Emitted when load is finished
- `failed`: Emitted when load fails

### Methods
- `get_progress`: Return the value contained on index 0 of `_progress` array
- `start_load`: Begin the load process
- `is_loaded`: Return true if the request has completed the loading

**Comment by Firemanarg** [comment]
# Modelling nº 3
```mermaid
---
title: Class Diagram - FDLoad (submodule of FDCore) and FDLoadRequest (subclass of FDLoad)
---
classDiagram
    class FDLoad {
        - abort_on_failure: bool
        - batch_size: int
        - _load_queue: Array[FDLoadRequest]
        - _batches: Array[FDLoadBatch]
        - _current_batch: int
        + signal started()
        + signal finished(has_failures: bool)
        + signal aborted()
        + signal progress_changed(progress: float)
        + abort(): void
        + add_to_queue(path: String, string_type: String, use_subthread: bool, retry_limit: int, cache_mode: CacheMode): bool
        + get_queue(): PackedStringArray
        + queue_remove_by_index(index: int): void
        + queue_remove_by_path(path: String): void
        + clear_queue(): void
        + start(): void
        + get_progress(): float
        + has_loaded(path: String): bool
        + get_loaded(path: String): Resource
        + _load_next_batch(): void
    }
    class FDLoadRequest {
        - path: String
        - use_subthread: bool
        - cache_mode: CacheMode
        - string_type: String
        - retry_limit: int
        - _progress: Array
        - _attempt_count: int
        + signal finished()
        + signal failed()
        + get_progress(): float
        + start_load(): void
        + is_loaded(): bool
    }
    class FDLoadBatch {
        - _requests: Array[FDRequest]
        - _has_aborted: bool
        + signal finished()
        + signal failed()
        + abort(): void
        + get_progress(): float
        + start_load(): void
        + is_loaded(): bool
    }
```
`Re-added manual abort method. Now processing will be made in batches, defined by class FDLoadBatch.`
`Removed retry failures. Now there will be a maximum retry limit, defined on each request.`
 
## Class: FDLoad
### Properties
- `abort_on_failure`: Flag to define if the loading will abort on first failure or if it will continue and storage the failures for a retry. If this flag is set, only the first failure will be added to failures queue
- `batch_size`: Requests count per batch
- `_load_queue`: Array containing all resource load requests
- `_batches`: Array containing load batches. Each batch contains up to `batch_size` requests
- `_current_batch`: Index of current batch being loaded

### Signals
- `started`: Emitted when load starts successfully
- `finished`: Emitted when load finishes successfully (without being aborted). First argument is a bool that is set if there was at least one failure during load
- `aborted`: Emitted when load finishes by abortion (if an error occurs and `abort_on_failure` flag is set, or if `abort()` is called)
- `progress_changed(progress)`: Emitted when progress change

### Methods
- `abort`: Aborts the load of unstarted batches. If there is a batch loading at the moment of the call, that batch will finish before abort next batches
- `add_to_queue(path, string_type, use_subthread, cache_mode)`: Create a load request from a resource path. Next arguments are optional, with default values `""`, `true` and `ResourceLoader.CacheMode.CACHE_MODE_REUSE` respectivelly for `string_type`, `use_subthread` and `cache_mode`
- `get_queue()`: Return a PackedStringArray containing load resource paths in queue to load
- `queue_remove_by_index(index)`: Removes a request by it's index
- `queue_remove_by_path(path)`: Removes a request by it's path
- `clear_queue()`: Delete all requests from `load_queue` and `failures_queue`
- `start()`: Begin the loading process of all resource requests from the load_queue. If a resource fails to load, it will be added to failures queue, and the load is aborted if the flag `abort_on_failure` is set. If failures queue contains any value before this method is called, it will be lost, since this method clears the failures queue when called
- `get_progress()`: Calculates the arithmetic average progress of all batches
- `has_loaded(path)`: Return `true` if the resource on `path` has been loaded previously
- `get_loaded(path)`: Return the loaded resource of given path. If the resource haven't been loaded previously, null is returned
- `_load_next_batch()`: Start the load of batch with index `_current_batch`. When finished, increments index

## Class: FDLoadRequest
### Properties
- `path`: Path of the resource to be loaded
- `use_subthread`: If false, loading process occurs in main thread, otherwise in subthread
- `cache_mode`: Cache mode of loading (check CacheMode from ResourceLoader)
- `string_type`: Optional property. If empty, the loaded resource will be handled as a generic Resource
- `retry_limit`: Limits of attempts of loading after failure. If `FDLoad.abort_on_failure` is `true`, this value will be ignored and no retry is executed
- `_progress`: Array to receive progress from ResourceLoader
- `_attempt_count`: Amount of attempts after failure. If this value is greater or equal to `retry_limit`, `failed` signal is emitted

### Signals
- `finished`: Emitted when load is finished
- `failed`: Emitted when load fails

### Methods
- `get_progress`: Return the value contained on index 0 of `_progress` array
- `start_load`: Begin the load process
- `is_loaded`: Return true if the request has completed the loading

## FDLoadBatch
### Properties
- `_requests`: Array containing all requests of this batch (a request is an instance of FDLoadRequest)
- `_has_aborted`: Flag to validate if an abortion occurred previously. The batch can only start loading if this flag is `false`

### Signals
- `finished`: Emitted when load is finished
- `failed`: Emitted when load fails

### Methods
- `abort`: Abort the processing of this batch and set the flag _has_aborted
- `get_progress`: Return the arithmetic average of all requests in the batch
- `start_load`: Begin the load process
- `is_loaded`: Return true if the request has completed the loading

**Comment by Firemanarg** [comment]
Resolved in #9
